File: ironpulse-frontend/src/types/community.ts
// ── Community types ──────────────────────────────────

export interface Author {
  id: string
  username: string
  display_name: string
  avatar_url: string | null
}

export type PostType = 'general' | 'pr' | 'workout_log' | 'physique' | 'challenge'

export interface Post {
  id: string
  user_id: string
  content: string
  post_type: PostType
  media_urls: string[] | null
  workout_session_id: string | null
  pr_id: string | null
  likes_count: number
  comments_count: number
  is_visible: boolean
  created_at: string
  updated_at: string
  author: Author | null
  is_liked_by_me: boolean
}

export interface Comment {
  id: string
  post_id: string
  user_id: string
  content: string
  created_at: string
  author: Author | null
}

export type ChallengeStatus = 'pending' | 'accepted' | 'declined' | 'completed' | 'expired'

export interface BattleChallenge {
  id: string
  challenger_id: string
  challenged_id: string
  exercise_id: string
  target_weight_kg: number | null
  target_reps: number | null
  status: ChallengeStatus
  expires_at: string | null
  created_at: string
}

export interface Message {
  id: string
  sender_id: string
  recipient_id: string
  content: string
  message_type: string
  metadata_json: Record<string, unknown> | null
  is_read: boolean
  created_at: string
}

export interface Conversation {
  partner_id: string
  partner_username: string
  partner_display_name: string
  partner_avatar_url: string | null
  last_message: Message
  unread_count: number
}

export type NotificationType = 'pr' | 'challenge' | 'social' | 'system'

export interface Notification {
  id: string
  user_id: string
  notification_type: NotificationType
  title: string
  body: string | null
  metadata_json: Record<string, unknown> | null
  is_read: boolean
  created_at: string
}

export interface PartnerCandidate {
  id: string
  username: string
  display_name: string
  avatar_url: string | null
  experience_level: string | null
  primary_goal: string | null
  training_days_per_week: number
  current_streak: number
  compatibility_score: number
}

export interface WebSocketEvent {
  event: string
  data: Record<string, unknown>
}

File: ironpulse-backend/app/schemas/community.py
"""
Community schemas — posts, comments, challenges, messages, notifications.
"""

import datetime
from pydantic import BaseModel, Field


# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
#  Author snippet (embedded in post / comment / message)
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
class AuthorBrief(BaseModel):
    id: str
    username: str
    display_name: str
    avatar_url: str | None = None


# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
#  Posts
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
class PostCreate(BaseModel):
    content: str
    post_type: str = "general"
    media_urls: list[str] = []
    workout_session_id: str | None = None
    pr_id: str | None = None


class PostOut(BaseModel):
    id: str
    user_id: str
    content: str
    post_type: str
    media_urls: list | None = None
    workout_session_id: str | None = None
    pr_id: str | None = None
    likes_count: int = 0
    comments_count: int = 0
    is_visible: bool = True
    created_at: datetime.datetime
    updated_at: datetime.datetime
    # Enriched fields
    author: AuthorBrief | None = None
    is_liked_by_me: bool = False

    model_config = {"from_attributes": True}


class PostFeedResponse(BaseModel):
    posts: list[PostOut]
    next_cursor: str | None = None


# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
#  Comments
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
class CommentCreate(BaseModel):
    content: str


class CommentOut(BaseModel):
    id: str
    post_id: str
    user_id: str
    content: str
    created_at: datetime.datetime
    author: AuthorBrief | None = None

    model_config = {"from_attributes": True}


# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
#  Battle Challenges
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
class ChallengeCreate(BaseModel):
    challenged_id: str
    exercise_id: str
    target_weight_kg: float | None = None
    target_reps: int | None = None
    expires_hours: int = 48


class ChallengeOut(BaseModel):
    id: str
    challenger_id: str
    challenged_id: str
    exercise_id: str
    target_weight_kg: float | None = None
    target_reps: int | None = None
    status: str
    expires_at: datetime.datetime | None = None
    created_at: datetime.datetime

    model_config = {"from_attributes": True}


# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
#  Messages / DMs
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
class MessageCreate(BaseModel):
    content: str
    message_type: str = "text"
    metadata_json: dict | None = None


class MessageOut(BaseModel):
    id: str
    sender_id: str
    recipient_id: str
    content: str
    message_type: str
    metadata_json: dict | None = None
    is_read: bool = False
    created_at: datetime.datetime

    model_config = {"from_attributes": True}


class ConversationOut(BaseModel):
    partner_id: str
    partner_username: str
    partner_display_name: str
    partner_avatar_url: str | None = None
    last_message: MessageOut
    unread_count: int = 0


# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
#  Notifications (Signals)
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
class NotificationOut(BaseModel):
    id: str
    user_id: str
    notification_type: str
    title: str
    body: str | None = None
    metadata_json: dict | None = None
    is_read: bool = False
    created_at: datetime.datetime

    model_config = {"from_attributes": True}


class SignalFeedResponse(BaseModel):
    notifications: list[NotificationOut]
    unread_count: int = 0


# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
#  Partner Discovery
# ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
class PartnerCandidate(BaseModel):
    id: str
    username: str
    display_name: str
    avatar_url: str | None = None
    experience_level: str | None = None
    primary_goal: str | None = None
    training_days_per_week: int = 3
    current_streak: int = 0
    compatibility_score: int = 0
